recipes/libcap/all/conanfile.py

```diff
@@ -0,0 +1,96 @@
+    _autotools_env = None
+
+    @property
+    def _source_subfolder(self):
+        return "source_subfolder"
+
+    def configure(self):
+        if self.settings.os != "Linux":
+            raise ConanInvalidConfiguration("Only Linux supported")
+        del self.settings.compiler.libcxx
+        del self.settings.compiler.cppstd
+
+    def source(self):
+        tools.get(**self.conan_data["sources"][self.version])
+        src_folder = "{}-{}".format(self.name, self.version)
+        os.rename(src_folder, self._source_subfolder)
+
+        make_rules = os.path.join(self._source_subfolder, "Make.Rules")
+        tools.replace_in_file(
+            make_rules,
+            "IPATH += -fPIC -I$(KERNEL_HEADERS) -I$(topdir)/libcap/include",
+            "IPATH += -I$(KERNEL_HEADERS) -I$(topdir)/libcap/include")
+
+    def _configure_autotools(self):
+        if self._autotools:
+            return self._autotools, self._autotools_env
+        self._autotools = AutoToolsBuildEnvironment(self)
+        self._autotools_env = self._autotools.vars
+        self._autotools_env["SHARED"] = \
+            "yes" if self.options.shared else "no"
+        self._autotools_env["PTHREADS"] = \
+            "yes" if self.options.psx_syscals else "no"
+        self._autotools_env["DESTDIR"] = self.package_folder
+        self._autotools_env["prefix"] = "/"
+        self._autotools_env["lib"] = "lib"
+        return self._autotools, self._autotools_env
+
+    def build(self):
+        with tools.chdir(os.path.join(self._source_subfolder, self.name)):
+            env_build, env_build_vars = self._configure_autotools()
+            env_build.make(vars=env_build_vars)
+
+    def package(self):
+        self.copy("License", dst="licenses", src=self._source_subfolder)
+
+        with tools.chdir(os.path.join(self._source_subfolder, self.name)):
+            env_build, env_build_vars = self._configure_autotools()
+
+            env_build.make(target="install-common-cap", vars=env_build_vars)
+            install_cap = ("install-shared-cap" if self.options.shared
+                           else "install-static-cap")
+            env_build.make(target=install_cap, vars=env_build_vars)
+
+            if self.options.psx_syscals:
+                env_build.make(target="install-common-psx",
+                               vars=env_build_vars)
+                install_psx = ("install-shared-psx" if self.options.shared
+                               else "install-static-psx")
+                env_build.make(target=install_psx, vars=env_build_vars)
+
+        tools.rmdir(os.path.join(self.package_folder, "lib", "pkgconfig"))
+
+    def package_info(self):
+        self.cpp_info.components["cap"].names["pkg_config"] = "libcap"
+        self.cpp_info.components["cap"].libs = ["cap"]
+        if self.options.psx_syscals:
+            self.cpp_info.components["psx"].names["pkg_config"] = "libpsx"
+            self.cpp_info.components["psx"].libs = ["psx"]
```

Comment:
Are the 2 components independent of eachother?
Does `psx` depend on `cap`?

Reply:
Not exactly. `psx` is only depend on `pthread`. `cap` has weak dependency on `psx`. [Linking of `psx` will override some `cap` functions](https://git.kernel.org/pub/scm/libs/libcap/libcap.git/tree/libcap/cap_proc.c#n77).

Reply:
So is this necessary:
`self.cpp_info.components["cap"].requires = ["psx"]`?
Just to be sure that psx is linked to first?

Reply:
Actually yes, we should link `psx` first. But it means we need `["psx"].requires = ["cap"]` not `["cap"].requires = ["psx"]` (I experimented with verbose making of test_package to be sure). And this will not allow to link `psx` without `cap` (I don't know who need it but anyway).
I'm not sure what is a proper way to resolve such kind of dependencies but let's add `psx` dependency on `cap`.

Reply:
I think this is incorrect, as it is now impossible to use `cap` without `psx`.
What can also be done:

```python
self.cpp_info.components["cap"].libs = ["cap"]
self.cpp_info.components["psx"].libs = ["psx", "cap"]
```

and don't add any requires.

Reply:
> I think this is incorrect, as it is now impossible to use `cap` without `psx`.

No, it is now impossible to use `psx` without `cap`.
If we will use `components["psx"].libs = ["psx", "cap"]` instead of requirements it doesn't change a lot. It still not allows to link `psx` without `cap`. Also, by some reasons, pkg_config-generator preserves the linkage order but cmake-generator doesn't. So without `requires` cmake generator breaks the resolving of weak references.

Actually that the recipe should do is:
* allow any package to link with `libcap`
* allow any package to link with `libpsx`
* allow any package to link with `libpsx libcap` in this order

I don't know how to do it, but can offer next solution:
* keep the current implementation with `components["psx"].requires = ["cap"]`
* don't specify any dependencies between `libcap` and `libpsx` and let package user to specify the correct order
* don't specify any dependencies between `libcap` and `libpsx` but add one another component `cap-with-psx`

I would prefer not to solve the problem with weak references overriding in the recipe, because original pkg_config files doesn't do it either. Also I don't know why the package should disallow to link both libraries without weak references overriding. It is a strange case but it is legal.

Reply:
I would look at the pc files:
- [`libcap.pc.in`](https://git.kernel.org/pub/scm/libs/libcap/libcap.git/tree/libcap/libcap.pc.in): only contains cap
- [`libpsx.pc.in`](https://git.kernel.org/pub/scm/libs/libcap/libcap.git/tree/libcap/libpsx.pc.in): only contains psx
And don't define any dependencies.
The consumers should take care of that.

Reply:
I will do the same then. No extra dependencies